Unpin pydantic version

We had to pin pydantic after 2.10 was release due to the following error:

```
E       pydantic.errors.PydanticUserError: `ErtConfig` is not fully defined; you should define `List`, then call `pydantic.dataclasses.rebuild_dataclass(ErtConfig)`.
E       
E       For further information visit https://errors.pydantic.dev/2.10/u/class-not-fully-defined

/Users/OYVEID/projects/ert/ert/src/ert/config/ert_config.py:525: PydanticUserError
```

to reproduce upgrade pydantic and run for example:
`tests/ert/unit_tests/config/test_ert_config.py`

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -53,7 +53,7 @@ dependencies = [
     "polars",
     "psutil",
     "pyarrow",  # extra dependency for pandas (parquet)
-    "pydantic > 2, < 2.10",
+    "pydantic > 2",
     "PyQt5",
     "python-dateutil",
     "python-multipart",  # extra dependency for fastapi

--- a/src/ert/config/ert_config.py
+++ b/src/ert/config/ert_config.py
@@ -20,7 +20,7 @@ from typing import (
 import polars
 from pydantic import ValidationError as PydanticValidationError
 from pydantic import field_validator
-from pydantic.dataclasses import dataclass
+from pydantic.dataclasses import dataclass, rebuild_dataclass
 
 from ert.plugins import ErtPluginManager
 from ert.substitutions import Substitutions
@@ -1166,3 +1166,7 @@ def _forward_model_step_from_config_file(
         )
     except OSError as err:
         raise ConfigValidationError.with_context(str(err), config_file) from err
+
+
+# Due to circular dependency in type annotations between ErtConfig -> WorkflowJob -> ErtScript -> ErtConfig
+rebuild_dataclass(ErtConfig)

